File: Diagrammatize.Design.Pattern/code/python/Proxy/A1/PrinterProxy.py
import threading

from Printable import Printable
from PrinterManager import PrinterManager

class PrinterProxy(Printable):
    def __init__(self, name, className):
        self.name = name
        self.className = className
        self.real = None
        self.lock = threading.Lock()

    def setPrinterName(self, name):
        with self.lock:
            if self.real is not None:
                self.real.setPrinterName(name)
            self.name = name

    def getPrinterName(self):
        return self.name

    def print(self, string):
        self.realize()
        self.real.print(string)

    def realize(self):
        with self.lock:
            if self.real is None:
                self.real = PrinterManager.newInstance(self.className)
                if self.real:
                    self.real.setPrinterName(self.name)
                else:
                    print("没有找到 " + className + " 类。")

File: Diagrammatize.Design.Patterns/code/cxx/Proxy/A1/PrinterProxy.cpp
#include "PrinterProxy.hpp"
#include "PrinterManager.hpp"
#include <iostream>

PrinterProxy::PrinterProxy(const std::string& name_, const std::string& className_): 
    name(name_), className(className_) {
}

void PrinterProxy::setPrinterName(const std::string& name_) {
    std::lock_guard<std::mutex> guard(lock);
    if (real != nullptr) {
        real->setPrinterName(name_);
    }
    name = name_;
}

std::string PrinterProxy::getPrinterName() {
    return name;
}

void PrinterProxy::print(const std::string& string) {
    realize();
    real->print(string);
}

void PrinterProxy::realize() {
    std::lock_guard<std::mutex> guard(lock);
    if (real == nullptr) {
        real = PrinterManager::newInstance(className);
        if (real) {
            real->setPrinterName(name);
        } else {
            std::cout << "没有找到 " << className << " 类。" << std::endl;
        }
    }
}
